src/VisualStudio/Core/Def/ProjectSystem/OpenTextBufferProvider.cs

```diff
@@ -0,0 +1,313 @@
+        {
+            await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            foreach (var (filePath, textBuffer, hierarchy) in EnumerateDocumentSet())
+            {
+                // We might or might not have seen this file be opened if it was opened between when we subscribed to the running document table and when
+                // we got scheduled to the UI thread.
+                if (!_monikerToTextBufferMap.ContainsKey(filePath))
+                {
+                    _monikerToTextBufferMap = _monikerToTextBufferMap.Add(filePath, textBuffer);
+
+                    foreach (var listener in _listeners)
+                        listener.OnOpenDocument(filePath, textBuffer, hierarchy);
+                }
+            }
+        }
+
+        public void AddListener(IOpenTextBufferEventListener listener) => ImmutableInterlocked.Update(ref _listeners, static (array, listener) => array.Add(listener), listener);
+        public void RemoveListener(IOpenTextBufferEventListener listener) => ImmutableInterlocked.Update(ref _listeners, static (array, listener) => array.Remove(listener), listener);
+
+        public int OnAfterFirstDocumentLock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
+            => VSConstants.E_NOTIMPL;
+
+        public int OnBeforeLastDocumentUnlock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
+        {
+            if (dwReadLocksRemaining + dwEditLocksRemaining == 0)
+            {
+                _foregroundAffinitization.AssertIsForeground();
+                if (_runningDocumentTable.IsDocumentInitialized(docCookie))
+                {
+                    var moniker = _runningDocumentTable.GetDocumentMoniker(docCookie);
+                    _monikerToTextBufferMap = _monikerToTextBufferMap.Remove(moniker);
+
+                    RaiseEventForEachListener(l => l.OnCloseDocument(moniker));
+                }
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterSave(uint docCookie)
+            => VSConstants.E_NOTIMPL;
+
+        public int OnAfterAttributeChange(uint docCookie, uint grfAttribs)
+            => VSConstants.E_NOTIMPL;
+
+        public int OnAfterAttributeChangeEx(uint docCookie, uint grfAttribs, IVsHierarchy pHierOld, uint itemidOld, string pszMkDocumentOld, IVsHierarchy pHierNew, uint itemidNew, string pszMkDocumentNew)
+        {
+            // Did we rename?
+            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_MkDocument) != 0)
+            {
+                _foregroundAffinitization.AssertIsForeground();
+                if (_runningDocumentTable.IsDocumentInitialized(docCookie))
+                {
+                    // We should already have a text buffer for this one
+                    if (_monikerToTextBufferMap.TryGetValue(pszMkDocumentOld, out var textBuffer))
```

Comment:
should this throw?

Reply:
We could, and then discover what esoteric case breaks it. Willing to be paranoid here.